panic: runtime error: index out of range

My coworker saw the below failure while building images on one of our longer lived buildkit daemons running v0.6.0 (I think!).

Hopefully the trace is enough, at the moment I don't have much more information about what was being built other than they restarted the container and were able to successfully process the same set of builds.

```
2019/08/04 18:12:06 http2: Transport failed to get client conn for cbir.clinc.ai:443: http2: no cached connection was available
time="2019-08-04T18:12:07Z" level=debug msg="fetch response received" digest="sha256:7360135413a6cc7ce9a6763f2b1606001b4fa9c5d5e2cefbf493e6fcd881d7fd" mediatype=application/vnd.docker.distribution.manifest.v2+json response.headers="map[Co
nnection:[keep-alive] Content-Length:[0] Date:[Sun, 04 Aug 2019 18:12:07 GMT] Docker-Content-Digest:[sha256:7360135413a6cc7ce9a6763f2b1606001b4fa9c5d5e2cefbf493e6fcd881d7fd] Docker-Distribution-Api-Version:[registry/2.0] Location:[https://cbir.clinc.ai/v2/clincapi-base/manifests/sha256:7360135413a6cc7ce9a6763f2b1606001b4fa9c5d5e2cefbf493e6fcd881d7fd] Server:[nginx/1.12.2] X-Content-Type-Options:[nosniff]]" size=4958 status="201 Created" url="https://cbir.clinc.ai/v2/clinc
api-base/manifests/faec986c0074154d22df529e39bf6fc0487368b63c6e306b16999e724b3abe0e"
time="2019-08-04T18:12:07Z" level=debug msg="remove snapshot" key=tvfp93mm4rmtyyabef37n1hfc snapshotter=overlayfs
time="2019-08-04T18:12:07Z" level=debug msg="remove content" key="sha256:354e65030fc8a8d0b85d523d8e1414c90601286711a4a12cc6e8c7d6ebcced02"
time="2019-08-04T18:12:07Z" level=debug msg="remove content" key="sha256:7360135413a6cc7ce9a6763f2b1606001b4fa9c5d5e2cefbf493e6fcd881d7fd"
time="2019-08-04T18:12:07Z" level=debug msg="schedule snapshotter cleanup" snapshotter=overlayfs
time="2019-08-04T18:12:07Z" level=debug msg="schedule content cleanup"
time="2019-08-04T18:12:07Z" level=debug msg="removed content" digest="sha256:354e65030fc8a8d0b85d523d8e1414c90601286711a4a12cc6e8c7d6ebcced02"
time="2019-08-04T18:12:07Z" level=debug msg="removed content" digest="sha256:7360135413a6cc7ce9a6763f2b1606001b4fa9c5d5e2cefbf493e6fcd881d7fd"
time="2019-08-04T18:12:07Z" level=debug msg="content garbage collected" d=3.757627ms
time="2019-08-04T18:12:07Z" level=debug msg="removed snapshot" key=buildkit/2899/tvfp93mm4rmtyyabef37n1hfc snapshotter=overlayfs
time="2019-08-04T18:12:07Z" level=debug msg="snapshot garbage collected" d=9.405294ms snapshotter=overlayfs
panic: runtime error: index out of range
goroutine 79237 [running]:
github.com/moby/buildkit/cache/remotecache/v1.sortConfig.func2(0x2, 0x1, 0x1)
        /src/cache/remotecache/v1/utils.go:74 +0x3f5
sort.insertionSort_func(0xc002d42ae8, 0xc000414520, 0x0, 0x6)
        /usr/local/go/src/sort/zfuncversion.go:12 +0xb1
sort.quickSort_func(0xc002d42ae8, 0xc000414520, 0x0, 0x6, 0x9)
        /usr/local/go/src/sort/zfuncversion.go:158 +0x1f3
sort.quickSort_func(0xc002d42ae8, 0xc000414520, 0x0, 0x1a, 0x9)
        /usr/local/go/src/sort/zfuncversion.go:145 +0xf8
sort.Slice(0xf4c2a0, 0xc000414500, 0xc002d42ae8)
        /usr/local/go/src/sort/slice.go:21 +0x129
github.com/moby/buildkit/cache/remotecache/v1.sortConfig(0xc000fd3410)
        /src/cache/remotecache/v1/utils.go:64 +0x586
github.com/moby/buildkit/cache/remotecache/v1.(*CacheChains).Marshal(0xc0003c2040, 0x1, 0x0, 0x0, 0xc0004aecc0)
        /src/cache/remotecache/v1/chains.go:83 +0x20d
github.com/moby/buildkit/cache/remotecache.export(0x12a4e60, 0xc00047c900, 0x1283780, 0xc000422b90, 0xc0003c2040, 0x0, 0xc0004cdd40, 0xc0006cf680)
        /src/cache/remotecache/export.go:70 +0x54
github.com/moby/buildkit/cache/remotecache.(*contentCacheExporter).Finalize(0xc001239740, 0x12a4e60, 0xc00047c900, 0x0, 0xc00047cb40, 0xc0004af100)
        /src/cache/remotecache/export.go:65 +0x55
github.com/moby/buildkit/solver/llbsolver.(*Solver).Solve.func3(0x12a4e60, 0xc00047c900, 0xc0004cdcc0, 0x0)
        /src/solver/llbsolver/solver.go:216 +0x190
github.com/moby/buildkit/solver/llbsolver.inVertexContext(0x12a4e60, 0xc00047c900, 0x1123ac2, 0xf, 0x0, 0x0, 0xc0004af498, 0x0, 0x0)
        /src/solver/llbsolver/solver.go:317 +0x233
github.com/moby/buildkit/solver/llbsolver.(*Solver).Solve(0xc00015d420, 0x12a4e60, 0xc0012394d0, 0xc000516060, 0x19, 0x0, 0xc00112cd40, 0xa, 0xc0012393e0, 0xc0003c2060, ...)
        /src/solver/llbsolver/solver.go:203 +0xf5a
github.com/moby/buildkit/control.(*Controller).Solve(0xc0002955f0, 0x12a4e60, 0xc001239470, 0xc0005ee900, 0x0, 0x0, 0x0)
        /src/control/control.go:276 +0x4b2
github.com/moby/buildkit/api/services/control._Control_Solve_Handler.func1(0x12a4e60, 0xc001239470, 0x10e6f60, 0xc0005ee900, 0x10cbe40, 0x1ce7ee8, 0x12a4e60, 0xc001239470)
        /src/api/services/control/control.pb.go:1364 +0x86
github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x12a4da0, 0xc0005e6bc0, 0x10e6f60, 0xc0005ee900, 0xc002d1de00, 0xc002d1de20, 0x0, 0x0, 0x1282ce0, 0xc0000950b0)
        /src/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:57 +0x2eb
main.unaryInterceptor.func1(0x12a4da0, 0xc0005e6bc0, 0x10e6f60, 0xc0005ee900, 0xc002d1de00, 0xc002d1de20, 0x0, 0x0, 0x0, 0x0)
        /src/cmd/buildkitd/main.go:509 +0x15f
github.com/moby/buildkit/api/services/control._Control_Solve_Handler(0x107a460, 0xc0002955f0, 0x12a4e60, 0xc001239380, 0xc0006d5180, 0xc00033cc40, 0x12a4e60, 0xc001239380, 0xc0007a9200, 0x231)
        /src/api/services/control/control.pb.go:1366 +0x158
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000088f00, 0x12b6d60, 0xc000bdb500, 0xc00010cb00, 0xc00055c090, 0x1c75378, 0x0, 0x0, 0x0)
        /src/vendor/google.golang.org/grpc/server.go:972 +0x470
google.golang.org/grpc.(*Server).handleStream(0xc000088f00, 0x12b6d60, 0xc000bdb500, 0xc00010cb00, 0x0)
        /src/vendor/google.golang.org/grpc/server.go:1252 +0xda6
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00112cd10, 0xc000088f00, 0x12b6d60, 0xc000bdb500, 0xc00010cb00)
        /src/vendor/google.golang.org/grpc/server.go:691 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /src/vendor/google.golang.org/grpc/server.go:689 +0xa1

```

--- a/cache/remotecache/v1/utils.go
+++ b/cache/remotecache/v1/utils.go
@@ -67,8 +67,8 @@ func sortConfig(cc *CacheConfig) {
 		if ri.Digest != rj.Digest {
 			return ri.Digest < rj.Digest
 		}
-		if len(ri.Inputs) != len(ri.Inputs) {
-			return len(ri.Inputs) < len(ri.Inputs)
+		if len(ri.Inputs) != len(rj.Inputs) {
+			return len(ri.Inputs) < len(rj.Inputs)
 		}
 		for i, inputs := range ri.Inputs {
 			if len(ri.Inputs[i]) != len(rj.Inputs[i]) {
@@ -76,7 +76,7 @@ func sortConfig(cc *CacheConfig) {
 			}
 			for j := range inputs {
 				if ri.Inputs[i][j].Selector != rj.Inputs[i][j].Selector {
-					return ri.Inputs[i][j].Selector != rj.Inputs[i][j].Selector
+					return ri.Inputs[i][j].Selector < rj.Inputs[i][j].Selector
 				}
 				return cc.Records[ri.Inputs[i][j].LinkIndex].Digest < cc.Records[rj.Inputs[i][j].LinkIndex].Digest
 			}

